<a href="https://colab.research.google.com/github/jyj1206/Python/blob/main/Python_everyday_problems.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 중간고사

Day 0
> 숫자를 입력 받아서 짝수인지 홀수인지 출력하는 프로그램을 작성하세요

In [ ]:
n = int(input("정수를 입력하세요 : "))
if n % 2 == 0:
    print(n, "은(는) 짝수입니다")
else:
    print(n, "은(는) 홀수입니다")

정수를 입력하세요 : 4
4 은(는) 짝수입니다


Day 1

> 사용자로부터 두 개의 정수를 입력으로 받습니다.
두 개 정수의 합, 정수의 차, 정수의 곱, 정수의 평균, 큰 수, 그리고 작은 수를 계산하여 화면에 출력하는 파이썬 프로그램을 작성하세요.

In [ ]:
num1, num2 = map(int, input("두 정수를 입력하세요").split())

print("두 정수의 합", num1 + num2)
print("두 정수의 차", num1 - num2 if num1 > num2 else num2 - num1)
print("두 정수의 곱", num1 * num2)
print("두 정수의 평균", (num1 + num2)/2)
print("두 정수 중 큰 수", num1 if num1 > num2 else num2)
print("두 정수 중 작은수", num2 if num1 < num2 else num1)

Day 2
> 사용자로부터 4자리의 정수를 입력으로 받아서 자리수의 합을 계산하는 프로그램을 작성하세요.
예를 들면 사용자가 1357을 입력한다면 1+3+5+7를 계산하고 그 결과를 출력하세요.

In [ ]:
n = int(input("4자리 수를 입력하세요."))

total = n % 10
n //= 10
total += n % 10
n //= 10
total += n % 10
n //= 10
total += n % 10

print(total)

Day 3
> BMI(Body Mass Index)는 체중(Kg)을 신장(m)의 제곱으로 나눈 값이다.
BMI는 체지방 축적을 잘 반영하기 때문에 비만도 판정에 많이 사용한다.
BMI 값이 18.50보다 작을 때는 저체중,
BMI 값이 18.50에서 24.99일 때는 표준,
BMI 값이 25 이상이고 29.99이하 일 때는 과체중,
BMI 값이 30 이상일때는 비만으로 판정한다.
사용자로부터 신장과 체중을 입력으로 받아서 BMI 값을 출력하는 프로그램을 작성하세요.

In [ ]:
height = float(input("신장을 입력하세요(cm 단위) : "))
weight = float(input("체중을 입력하세요(kg 단위) : "))
BMI = weight / ((height/100) ** 2)
if BMI >= 30:
  print("비만")
elif BMI >= 25:
  print("과체중")
elif BMI >= 18.5:
  print("표준")
else:
  print("저체중")

Day 4
> 사용자로부터 2개의 정수를 입력으로 받습니다.
두 수가 같으면 1을, 다르면 -1을 출력하는 프로그램을 작성하세요.

In [ ]:
a, b = map(int, input("두 정수를 입력하세요 : ").split())

if a==b:
  print("1")
else:
  print("-1")

Day 5
> 각(angle)에서 0도 초과 90도 미만은 예각, 90도는 직각, 90도 초과 180도 미만은 둔각 180도는 평각으로 분류합니다.
사용자가 각을 입력했을 때, 예각이면 1, 직각이면 2, 둔각이면 3, 평각이면 4를 출력하는 프로그램을 작성하세요.
사용자의 입력은 0부터 180이라고 가정하자.

In [ ]:
angle = int(input("각을 입력하세요 : "))
if angle == 180:
  print("평각")
elif angle > 90:
  print("둔각")
elif angle == 90:
  print("직각")
else:
  print("예각")

Day 6
> 사용자로부터 2개의 정수를 입력으로 받습니다.
각각의 정수를 num1과 num2라고 한다.
num1이 num2의 배수일 경우에는 "num1은 num2의 배수입니다."라고 출력한다.
num1이 num2의 배수가 아닐 경우에는 "num1은 num2의 배수가 아닙니다."라고 출력한다.

In [ ]:
num1, num2 = map(int, input("2개의 정수를 입력하세요(단, 2번째 입력하는 수는 0이 아님) : ").split())
if (num1 % num2 == 0) :
    print(f"{num1}은(는) {num2}의 배수입니다.")
else :
    print(f"{num1}은(는) {num2}의 배수가 아닙니다.")

Day 7
> 새로 생 피자 가게는 피자를 일곱 조각으로 잘라 줍니다.
사용자로부터 피자를 나눠 먹을 사람의 수 n을 입력으로 받는다.
모든 사람이 피자를 한 조각 이상 먹기 위해 필요한 피자의 수를 계산하여 출력하는 프로그램을 작성하세요.

In [ ]:
from math import ceil
n = int(input("피자를 나눠 먹을 사람의 수 : "))
print(ceil(n/7))

Day 8
> 춘식이는 추운 겨울에도 아이스 아메리카노만 마시는 소위 말하는 얼죽아입니다.
아이스 아메리카노는 한 잔에 4,500원입니다.
춘식이가 가지고 있는 돈 money를 사용자로부터 입력으로 받습니다.
춘식이가 최대로 마실 수 있는 아이스 아메리카노의 잔 수와 남는 돈을 순서대로 담은 리스트를 출력하는 프로그램을 작성하세요.

In [ ]:
money = int(input("춘식이가 가지고 있는 돈을 입력하세요 : "))
print([money // 4500, money % 4500])

Day 9
> 정수를 담은 임의의 리스트 num_list가 있습니다.
가장 첫 번째 원소를 1번 원소라고 할 때,
홀수 번째 원소들의 합과 짝수 번째 원소들의 합을 계산합니다.
두 합 중에서 큰 값을 출력하는 프로그램을 작성하세요.

In [ ]:
num_list = list(range(1, 11)) # 임의의 리스트

# 홀수번째 원소 합
odd_sum = sum([num_list[i] for i in range(1, len(num_list), 2)])

# 짝수번째 원소 합
even_sum = sum([num_list[i] for i in range(0, len(num_list), 2)])

# 두 합 중 큰 값 출력
print(max(odd_sum, even_sum))

Day 10
> 10이하의 자연수에서 3 또는 5의 배수는, 3, 5, 6, 9이고 이 숫자들의 합은 23이다.
1000이하의 자연수에서 3 또는 5의 배수인 숫자들의 합을 구하는 프로그램을 작성하세요.

In [ ]:
total = 0
for n in range(1, 1001):
  if n % 3 == 0 or n % 5 == 0:
    total += n
print(total)

Day 11
> 피보나치 수열(Fibonacci sequence)은 각 항이 앞의 두 항의 합으로 이루어진 수열입니다.
사용자로부터 양의 정수 n을 입력으로 받아서 피보나치 수열의 n번째 항의 값을 출력하는 프로그램을 작성하세요.
단, 반복문을 이용하여 프로그램을 작성하세요.
피보나치 수열에 대한 자세한 사항은 아래의 내용을 참고하세요.

>피보나치 수열의 정의:
   - 첫 번째 항: 0
   - 두 번째 항: 1
   - 세 번째 항부터는 앞의 두 항의 합으로 계산됩니다.

> 즉, 피보나치 수열은 다음과 같습니다.
0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, ...
>
> - 첫 번째 항 F(0)=0
> - 두 번째 항 F(1)=1
> - 세 번째 항 F(2)=F(1)+F(0)=1+0=1
> - 네 번째 항 F(3)=F(2)+F(1)=1+1=2
> - 다섯 번째 항 F(4)=F(3)+F(2)=2+1=3

>피보나치 수열을 수식으로 나타내면 다음과 같습니다.
>- F(n)=F(n−1)+F(n−2) (단, n≥2)

In [ ]:
n = int(input("피보나치 수의 몇 번째 항인지 입력하세요.(단 n은 1부터 시작) : "))
fibo = [0] * (n + 1)
fibo[2] = 1
if n == 1 or  n == 2:
  print(fibo[n])
else:
  for i in range(3, n + 1):
    fibo[i] = fibo[i - 1] + fibo[i - 2]
  print(fibo[n])

Day 12
> 공백(white space)을 다수 포함한 문자열이 주어졌다.
공백을 모두 제거한 새로운 문자열을 만들어 보고자 한다.
원래 문자열과 새로운 문자열을 출력하는 프로그램을 작성하세요.

In [ ]:
s = input("공백을 다수 포함한 문자열 입력 : ")
print("원래 문자열 :", s)
print("새로운 문자열 :", s.replace(' ', ''))

공백을 다수 포함한 문자열 입력 : 아 아 아아
원래 문자열 : 아 아 아아
새로운 문자열 : 아아아아


Day 13
> <이상한 변호사 우영우?를 보셨나요?
주인공 우영우는 자기 소개를 할 때,
"똑바로 읽어도 거꾸로 읽어도 우영우입니다.
기러기, 인도인, 토마토, 스위스, 별똥별, 우영우"라고 말합니다.
똑바로 읽어도 거꾸로 읽어도 같은 문자열을 회문(回文)이라고 합니다.
영어로는 palindrome 이라고 합니다.
사용자로부터 입력을 받은 문자열이 회문(回文) / palindrome 인지 아닌지 판별하는 프로그램을 작성하세요.

In [ ]:
sentence = input("문자열을 입력하세요 : ")
if sentence == sentence[::-1]:
  print("회문입니다.")
else:
  print("회문이 아닙니다.")

문자열을 입력하세요 : 토마토
회문입니다.


Day 14
> 사용자로부터 두 개의 문자열을 입력으로 받습니다.
첫 번째 문자열 속에 두 번째 문자열이 존재하는지 여부를 판별하는 프로그램을 작성하세요.

In [ ]:
first_sentence = input("첫 번째 문자열을 입력하세요 : ")
second_sentence = input("두 번째 문자열을 입력하세요 : ")
if second_sentence in first_sentence:
  print("첫 번째 문자열 속에 두 번째 문자열이 존재합니다.")
else:
  print("첫 번째 문자열 속에 두 번째 문자열이 존재하지 않습니다.")

첫 번째 문자열을 입력하세요 : 사과 많이 많이
두 번째 문자열을 입력하세요 :  많이
첫 번째 문자열 속에 두 번째 문자열이 존재합니다.


Day 15
> "에라토스테네스(Eratosthenes)의 체"가 무엇일까요?
소수(Prime Number)를 구하는 방법 중에 하나입니다.
자세한 내용은 아래의 링크를 참조하시기 바랍니다.
"에라토스테네스의 체"를 활용하여, 1부터 10,000까지의 소수를 구해볼까요?
총 소수의 갯수와, 소수들을 나열하는 프로그램을 작성하세요.

In [ ]:
is_prime_list = [False, False] + [True] * 9999

for i in range(2, int(len(is_prime_list) ** 0.5) + 1):
  if is_prime_list[i]:
    for j in range(i + i, len(is_prime_list),i):
      is_prime_list[j] = False

print(sum(is_prime_list))
print([i for i, is_prime in enumerate(is_prime_list) if is_prime == True])

1229
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 121

Day 16
> 함수의 파라미터로 두 개의 문자열을 받습니다.
두 개의 문자열이 anagram이면 True를 반환하고, 아니면 False를 반환합니다.

> 사용자로부터 두 개의 문자열을 받아서, anagram 여부를 판별하는 프로그램을 작성하세요.

> Anagram의 예를 들면 다음과 같습니다.
1. listen - silent
2. race - care
3. eat - tea

In [ ]:
def is_anagram(str1, str2):
  # 대 소문자 구분 없음
  str1 = str1.lower()
  str2 = str2.lower()

  # 공백 제거
  str1 = str1.replace(" ", "")
  str2 = str2.replace(" ", "")

  # 두 문자열의 길이가 같지 않으면
  if len(str1) != len(str2):
    return False
  else :
    alpa = [0] * 26

    # 첫번째 문자열에 각 알파벳 갯수 저장
    for ch in str1:
      alpa[ord(ch) - ord('a')] += 1

    # 두번째 문자열과 비교
    for ch in str2:
      index = ord(ch) - ord('a')
      if alpa[index] <= 0:
        return False
      alpa[index] -= 1

    return True

string1 = input("첫 번째 문자열 : ")
string2 = input("첫 번째 문자열 : ")

if is_anagram(string1, string2):
  print("두 문자열은 Anagram입니다.")
else:
  print("두 문자열은 Anagram이 아닙니다.")

첫 번째 문자열 : Moon starer
첫 번째 문자열 : Astronomer
두 문자열은 Anagram입니다.


Day 17
> 함수를 만들어볼까요?
매개변수로 문자열을 받습니다.
문자열에 포함된 모음을 제거한 새로운 문자열을 반환합니다.
문자열은 영어 기준입니다.

> 사용자로부터 입력으로 받은 문자열을 인자로 사용하여 위의 함수를 호출합니다.
직접 만든 함수가 제대로 동작하는지 테스트 결과를 살펴보세요.

In [ ]:
def remove_vowels(s):
  vowels = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']

  new_s = ''
  for ch in s:
    if ch in vowels:
      continue
    new_s += ch
  return new_s

str_ = input("문자열을 입력하세요 : ")
print(f"모음을 제거한 새로운 문자열은 {remove_vowels(str_)} 입니다.")

문자열을 입력하세요 : string
모음을 제거한 새로운 문자열은 strng 입니다.


Day 18
> 함수를 만들어볼까요?
매개변수로 받은 문자열에서 가장 긴 단어를 찾아내서 반환해주는 함수입니다.

> 사용자로부터 한 문장(영어)을 문자열로 입력으로 받습니다.
그 문자열을 인자로 사용하여 만든 함수를 호출합니다.
본인이 만든 함수가 제대로 동작하는지 테스트해볼까요?

In [ ]:
def longest_word(s):
  words = s.split()
  longest = ''
  for word in words:
    if len(word) > len(longest):
      longest = word
  return longest

sentence = input("영어 문장을 입력하세요 : ")
print(f"문자열에서 가장 긴 단어는 {longest_word(sentence)}입니다.")

영어 문장을 입력하세요 : I love you
문자열에서 가장 긴 단어는 love입니다.


Day 19
> 회문문자(palindrome)을 판단하는 파이썬의 간결한 표현은 다음과 같습니다.
```python
is_palindrome = lambda s: s == s[::-1]
```
>  회문문자가 있다면 회문숫자가 있겠죠?
> 회문숫자는 앞뒤가 대칭인 숫자, 예를 들면, 9009입니다.
> 9009는 두 자리 숫자 두 개의 곱(91*99)을 통해서 만들 수 있는 가장 큰 회문숫자입니다.
> 세 자리의 숫자 두 개의 곱을 통해서 만들 수 있는 가장 큰 회문숫자를 찾는 프로그램을 만들어볼까요?
> 그 회문숫자를 구성하는 세 자리의 숫자 두 개는 무엇일까요?

In [ ]:
def is_palindrome(s):
  return s == s[::-1]

max_palindrome = 0
for i in range(999, 99, -1):
  for j in range(i, 99, -1):
    if is_palindrome(str(i * j)):
      if i * j > max_palindrome:
        max_palindrome = i * j
        max_pair = (i, j)

print(f"세자리 숫자의 두 개의 곱을 통해서 만들 수 있는 가장 큰 회문숫자는 {max_palindrome}입니다.")
print(f"그 회문 숫자를 구성하는 두 개는 {max_pair[0]}(와)과 {max_pair[1]}입니다.")

세자리 숫자의 두 개의 곱을 통해서 만들 수 있는 가장 큰 회문숫자는 906609입니다.
그 회문 숫자를 구성하는 두 개는 993(와)과 913입니다.


Day 20
> 반복문과 break 명령어를 활용하여 아래와 같은 결과를 출력하는 프로그램을 작성하세요.

```
#
# 1
# 1 2
# 1 2 3
# 1 2 3 4
# 1 2 3 4 5
```

In [ ]:
for i in range(6):
  print("#", end = ' ')
  for j in range(1, 6):
    if j > i:
      break
    print(j, end= ' ')
  print()

# 
# 1 
# 1 2 
# 1 2 3 
# 1 2 3 4 
# 1 2 3 4 5 


Day 21
> 정수 10개를 담고 있는 리스트가 있습니다.
값이 작은 정수부터 크기 순서로 차례대로 정렬한 리스트를 만드는 프로그램을 작성하세요.
Python 내부함수를 이용하여 리스트를 정렬하는 방법은 권하지 않습니다.

In [ ]:
def selection_sort(numbers):
  for i in range(9):
    min_index = i
    for j in range(i+1, 10):
      if numbers[min_index] > numbers[j]:
        min_index = j
    numbers[i], numbers[min_index] = numbers[min_index], numbers[i]

numbers = [9, 5, 4, 2, 3, 7, 8, 1, 10, 6]
selection_sort(numbers)
print(numbers)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


Day 22
> 영어 문장이 주어졌을 때, 특정 단어가 그 문장의 몇 번째 단어인지 확인하는 프로그램을 작성하세요.
예를 들면, "If music be the food of love, play on." 문장에서 "play"는 8번째 단어입니다.

In [ ]:
def find_word(sentence, word):
  replaced_sentence = sentence.replace("."," ").replace(",", " ").replace("!", " ").replace("?", " ") # 구두점 제거

  location = [w for w in replaced_sentence.lower().split()].index(word.lower()) + 1
  print(f'"{sentence}"문장에서 "{word}"는 {location}번쨰 단어입니다.')

sentence = input("영어 문장을 입력하세요 : ")
word = input("찾고자하는 단어를 입력하세요 : ")
find_word(sentence, word)

영어 문장을 입력하세요 : If music be the food of love, play on.
찾고자하는 단어를 입력하세요 : play
"If music be the food of love, play on."문장에서 "play"는 8번쨰 단어입니다.


Day 23
> "To be, or not to be, that is the question" 라는 문장을 들어보셨나요?
이 문장에는 단어 "be"가 두 차례 나타납니다.
사용자 입력으로 받은 문장에서 특정 단어가 몇 차례 나타나는 지를 알려주는 프로그램을 작성해보세요.

In [ ]:
def return_count_word(sentence, word):
  replaced_sentence = sentence.replace("."," ").replace(",", " ").replace("!", " ").replace("?", " ") # 구두점 제거

  words = replaced_sentence.lower().split()
  count_word = words.count(word.lower())
  return count_word

sentence = input("영어 문장을 입력하세요 : ")
word = input("찾고자하는 단어를 입력하세요 : ")
print(f'"{sentence}"문장에서 "{word}"는 {return_count_word(sentence, word)}번 나타납니다.')

영어 문장을 입력하세요 : To be, or not to be, that is the question
찾고자하는 단어를 입력하세요 : be
"To be, or not to be, that is the question"문장에서 "be"는 2번 나타납니다.


Day 24
> 영문자 a를 1, b를 2, c를 3이라고 하고, 이후에 모든 영문자에 숫자를 차례대로 부여하자.
사용자로부터 영어 단어를 하나 입력 받습니다.
그 단어에 포함된 모든 영어 단어에 해당하는 숫자의 값을 모두 더한다.
영어 단어와 함께 그 숫자들의 합을 출력하는 프로그램을 작성하세요.
이 프로그램을 작성할 때, 내장함수(built-in) 함수 ord( )를 참고하세요.

In [ ]:
word = input("영어 단어 하나를 입력하세요 (소문자) : ")
print(sum(map(lambda x : ord(x) - ord('a') + 1 , word)))

영어 단어 하나를 입력하세요 (소문자) : apple
50


Day 25
> 모든 자연수 x에 대해서 현재 값이 x이면 x가 짝수일 때는 2로 나누고, x가 홀수일 때는 3 * x + 1로 바꾸는 계산을 계속해서 반복하면 언젠가는 반드시 x가 1이 되는지 묻는 문제를 콜라츠 문제라고 부릅니다.
그리고 위 과정에서 거쳐간 모든 수를 기록한 수열을 콜라츠 수열이라고 부릅니다.
계산 결과 1,000 보다 작거나 같은 수에 대해서는 전부 언젠가 1에 도달한다는 것이 알려져 있습니다.

>임의의 1,000 보다 작거나 같은 양의 정수 n이 주어질 때 초기값이 n인 콜라츠 수열을 반환하는 함수 collatz를 완성해 주세요.
collatz 함수가 제대로 작동하는지 테스트하는 코드를 작성하세요.

>입력 n : 10
출력 result : [10, 5, 16, 8, 4, 2, 1]

In [ ]:
def return_collatz(n):
  result = [n]
  while n != 1:
    if n % 2 == 0:
      n //= 2
    else:
      n = 3 * n + 1
    result.append(n)

  return result

n = int(input("입력 n : "))
print("출력 result :", return_collatz(n))

입력 n : 3
출력 result : [3, 10, 5, 16, 8, 4, 2, 1]


Day 26
> 약수의 개수가 세 개 이상인 수를 합성수라고 합니다.

> 자연수 n이 매개변수로 주어질 때 n이하의 합성수의 개수를 반환하도록 함수를 구현하세요.

> 그 함수가 제대로 동작하는지 테스트하는 코드를 작성하세요.
사용자로부터 n을 입력으로 받고, 합성수의 개수를 출력합니다.
예를 들면, 15 이하 합성수는 4, 6, 8, 9, 10, 12, 14, 15 로 8개입니다.
따라서 8을 출력합니다

In [ ]:
def count_composite_number(n):
  count = 0
  # 소수가 아니면 합성수이므로, 단 1은 약수의 개수가 1개이므로 제외
  for i in range(2, n + 1):
    if is_prime(i):
      continue
    count += 1
  return count


def is_prime(n):
  if n == 1:
    return False
  for i in range(2, int(n ** 0.5) + 1):
    if n % i == 0:
      return False
  return True

n = int(input("자연수 n을 입력하세요 : "))
print(f"{n} 이하 합성수의 개수는 {count_composite_number(n)}개 입니다.")

자연수 n을 입력하세요 : 15
15 이하 합성수의 개수는 8개 입니다.


Day 27
> 재귀함수(recursive function)를 이용해서 문제를 해결할 수 있는 방법을 설명해드렸습니다.
정수 n과 m 사이의 모든 정수의 합을 구하는 문제도 재귀를 사용하여 풀 수도 있습니다.
조건은 n은 m보다 작은 정수라는 것입니다.

In [ ]:
def sum_numbers(n, m):
  if n == m:
    return n
  else:
    return n + sum_numbers(n + 1, m)

n = int(input("정수 n을 입력하세요 : "))
m = int(input("정수 m을 입력하세요 : "))
print(f"{n}과 {m}사이의 모든 정수의 합은 {sum_numbers(n, m)}입니다.")

정수 n을 입력하세요 : 4
정수 m을 입력하세요 : 5
4과 5사이의 모든 정수의 합은 9입니다.


Day 28 시험 주 생략

# 기말고사

Day 29
> 양의 정수 n이 매개변수로 주어질 때, n이 홀수라면 n 이하의 홀수인 모든 양의 정수의 합을 반환하고 n이 짝수라면 n 이하의 짝수인 모든 양의 정수의 제곱의 합을 반환하는 함수를 작성해 주세요. 그 함수가 제대로 작동하는지 테스트하는 프로그램을 작성하세요.

In [ ]:
def sum_odd_or_even(n):
  if n % 2 == 0:
    return sum([i ** 2 for i in range(2, n + 1, 2)])
  else:
    return sum([i for i in range(1, n + 1, 2)])

n = int(input("양의 정수 n을 입력하세요 : "))
print(f"결과는 {sum_odd_or_even(n)}입니다.")

양의 정수 n을 입력하세요 : 10
결과는 220입니다.


Day 30
> 2자리 이상의 정수 number가 사용자의 입력으로 주어집니다.  이 수를 2자리씩 자른 뒤, 자른 수를 모두 더해서 그 합을 출력하는 프로그램을 작성하세요.

>입출력 예제
>
>입력 : 4859
>
>출력 : 107
>
>입력된 수를 2자리씩 나눠 합치면 다음과 같습니다.
>
>48 + 59 = 107

In [ ]:
n = input("2자리 이상의 정수를 입력하세요 : ")
total = 0
for i in range(0, len(n), 2):
  total += int(n[i:i+2])
print(total)

2자리 이상의 정수를 입력하세요 : 4859
107


Day 31
> 자연수 n을 뒤집어 각 자리 숫자를 원소로 가지는 배열 형태로 반환하는 함수를 작성해주세요. 예를 들어 매개변수 n이 12345이면 [5,4,3,2,1]을 반환합니다. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.

In [ ]:
n = int(input("자연수 n을 입력하세요 : "))
print(list(map(int, list(str(n)[::-1]))))

자연수 n을 입력하세요 : 12345
[5, 4, 3, 2, 1]


Day 32
> 매개변수로 주어진 정수를 저장한 리스트 numbers에서 가장 작은 수를 제거한 리스트를 반환하는 함수를 작성해주세요. 단, 반환하려는 리스트가 빈 리스트인 경우엔 리스트에 -1을 채워 반환하세요. 예를 들어 numbers이 [4,3,2,1]인 경우는 [4,3,2]를 리턴 하고, [10]면 [-1]을 반환합니다. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.

In [ ]:
numbers = list(map(int, input("여러 개 수를 공백을 포함하여 입력하세요 : ").split()))

min_n = min(numbers)
numbers.remove(min_n)

if not numbers:
  numbers.append(-1)

print(numbers)

여러 개 수를 공백을 포함하여 입력하세요 : 10
[-1]


Day 33
> 정수 리스트 numbers가 매개변수로 주어집니다. numbers에서 서로 다른 인덱스에 있는 두 개의 수를 뽑아 더해서 만들 수 있는 모든 수를 리스트에 오름차순으로 담아 반환하도록 함수를 완성해주세요. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.
다음과 같은 제한사항이 있습니다.
> 1. numbers의 길이는 2 이상 100 이하입니다.
> 2. numbers의 모든 수는 0 이상 100 이하입니다.

In [ ]:
numbers = list(map(int, input("정수 리스트를 입력하세요 : ").split()))

result = []
for i in range(len(numbers)):
  for j in range(i + 1, len(numbers)):
    result.append(numbers[i] + numbers[j])

print(sorted(list(set(result))))

정수 리스트를 입력하세요 : 1 2 3 4 5 6 7 8 9 10
[3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]


Day 33
> 특정 날짜를 입력하면 그 날짜까지 남은 날이 얼마인지 알려주는 파이썬 코드를 작성해보세요.

In [ ]:
from datetime import datetime, timedelta, timezone

input_date = input("특정 날짜를 입력하세요 (YYYY/MM/DD) : ")
date = datetime.strptime(input_date, "%Y/%m/%d")

# 한국 표준시로 변경
korea_timezone = timezone(timedelta(hours=9))
korea_now = datetime.now(korea_timezone)

date = date.replace(tzinfo=korea_timezone)

# 입력한 날짜와 현재 날짜의 시간 부분을 모두 자정으로 맞추기
date = date.replace(tzinfo=korea_timezone, hour=0, minute=0, second=0, microsecond=0)
korea_now = korea_now.replace(hour=0, minute=0, second=0, microsecond=0)

diff = date - korea_now

print(f"{input_date}까지 남은 날은 {diff.days}일 입니다.")

특정 날짜를 입력하세요 (YYYY/MM/DD) : 2024/12/06
2024-12-06 00:00:00+09:00 2024-10-31 00:00:00+09:00
2024/12/06까지 남은 날은 36일 입니다.


Day 34
> 어떤 문장의 각 알파벳을 일정한 거리만큼 밀어서 다른 알파벳으로 바꾸는 암호화 방식을 "시저 암호(Caesar cipher)"라고 합니다. 예를 들어 "AB"는 1만큼 밀면 "BC"가 되고, 3만큼 밀면 "DE"가 됩니다. "z"는 1만큼 밀면 "a"가 됩니다. 매개변수로 문자열 s와 거리 n을 입력으로 받아 s를 n만큼 민 암호문을 만드는 함수를 완성해주세요. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.

> 다음과 같은 제한 조건을 가집니다.
- 공백은 아무리 밀어도 공백입니다.
- s는 알파벳 소문자, 대문자, 공백으로만 이루어져 있습니다.
- s의 길이는 8000 이하입니다.
- n은 1 이상, 25 이하인 자연수입니다.

In [ ]:
def caesar_cipher(s, n):
  result = ''
  for ch in s:
    if ch.isalpha():
      if ch.isupper(): # 대문자인 경우
        result += chr((ord(ch) - ord('A') + n) % 26 + ord('A'))
      else: # 소문자인 경우
        result += chr((ord(ch) - ord('a') + n) % 26 + ord('a'))
    else: # 공백인 경우
      result += ' '
  return result


s = input("문자열을 입력하세요 : ")
n = int(input("거리 n을 입력하세요 : "))
print(caesar_cipher(s, n))

문자열을 입력하세요 : ABC DEF
거리 n을 입력하세요 : 3
DEF GHI


Day 35
>  자연수 n이 매개변수로 주어집니다. n을 x로 나눈 나머지가 1이 되도록 하는 가장 작은 자연수 x를 반환하는 함수를 완성해주세요. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.

In [ ]:
def find_x(n):
  if n == 1:
      return 2  # 1을 2로 나누면 나머지가 1
  if n == 2:
      return -1  # 조건을 만족하는 x가 존재하지 않음

  # n%x == 1을 푸는 문제는 n-1의 약수를 찾는 문제와 동일
  for x in range(2, int((n - 1) ** 0.5) + 1):
    if (n - 1) % x == 0:
        return x
  # 약수가 없는 경우 n-1이 답
  return n - 1

n = int(input("자연수 n을 입력하세요 : "))
result = find_x(n)
if result != -1:
    print(result)
else:
    print("조건을 만족하는 x가 존재하지 않습니다.")

자연수 n을 입력하세요 : 12
11


Day 36
> 문자열 s는 한 개 이상의 단어로 구성되어 있습니다. 각 단어는 하나 이상의 공백문자로 구분되어 있습니다. 각 단어의 짝수번째 알파벳은 대문자로, 홀수번째 알파벳은 소문자로 바꾼 문자열을 리턴하는 함수를 완성하세요. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.
>
> 제한 사항
- 문자열 전체의 짝/홀수 인덱스가 아니라, 단어(공백을 기준)별로 짝/홀수 인덱스를 판단해야합니다.
- 첫 번째 글자는 0번째 인덱스로 보아 짝수번째 알파벳으로 처리해야 합니다.


In [ ]:
def transform_string(s):
    result = ''
    for word in s.split():
        for i in range(len(word)):
            if i % 2 == 0:
                result += word[i].upper()
            else:
                result += word[i].lower()
        result += ' '  # 각 단어 뒤에 공백 추가

    return result.strip()  # 마지막 공백 제거 후 반환

s = input("문자열을 입력하세요: ")
result = transform_string(s)
print(result)

문자열을 입력하세요: I love you
I LoVe YoU


Day 37
> 매개변수로 정수 n을 입력으로 받아 n의 약수를 모두 더한 값을 반환하는 함수를 완성해주세요.  이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.

In [ ]:
def sum_primes(n):
  total = 0
  for i in range(1, int(n ** 0.5) + 1):
    if n % i == 0:
      total += i
      if i != n // i:
        total += n // i
  return total

n = int(input("자연수 n을 입력하세요 : "))
print(sum_primes(n))

자연수 n을 입력하세요 : 10
18


Day 38
> 정수로 이루어진 배열 numbers가 있습니다. 배열 의 각 원소들에 대해 자신보다 뒤에 있는 숫자 중에서 자신보다 크면서 가장 가까이 있는 수를 뒷 큰수라고 합니다.
>
> 정수 배열 numbers가 매개변수로 주어질 때, 모든 원소에 대한 뒷 큰수들을 차례로 담은 배열을 반환하도록 함수를 완성해주세요. 단, 뒷 큰수가 존재하지 않는 원소는 -1을 담습니다.
이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.
제한사항은 다음과 같습니다.
>
> 4 ≤ numbers의 길이 ≤ 1,000,000
>
> 1 ≤ numbers[i] ≤ 1,000,000
>
> 입출력의 예는 다음과 같습니다.
>
> numbers             result
>
> [2, 3, 3, 5]           [3, 5, 5, -1]
>
> [9, 1, 5, 3, 6, 2]   [-1, 5, 6, 6, -1, -1]
>
> 입출력 예 #1
2의 뒷 큰수는 3입니다. 첫 번째 3의 뒷 큰수는 5입니다. 두 번째 3 또한 마찬가지입니다. 5는 뒷 큰수가 없으므로 -1입니다. 위 수들을 차례대로 배열에 담으면 [3, 5, 5, -1]이 됩니다.
>
> 입출력 예 #2
9는 뒷 큰수가 없으므로 -1입니다. 1의 뒷 큰수는 5이며, 5와 3의 뒷 큰수는 6입니다. 6과 2는 뒷 큰수가 없으므로 -1입니다. 위 수들을 차례대로 배열에 담으면 [-1, 5, 6, 6, -1, -1]이 됩니다

In [ ]:
def find_next_greater_element(numbers):
  result = []
  for i in range(len(numbers)):
    for j in range(i + 1, len(numbers)):
      if numbers[i] < numbers[j]:
        result.append(numbers[j])
        break
    else:
      result.append(-1)
  return result

numbers = list(map(int, input("정수 리스트를 입력하세요 : ").split()))
print(find_next_greater_element(numbers))

정수 리스트를 입력하세요 : 9 1 5 3 6 2
[-1, 5, 6, 6, -1, -1]


Day 39
> 0 또는 양의 정수가 여러 개 주어졌을 때, 정수를 이어 붙여 만들 수 있는 가장 큰 수를 찾아주세요. 예를 들어, 주어진 정수가 [6, 10, 2]라면 [6102, 6210, 1062, 1026, 2610, 2106]를 만들 수 있고, 이중 가장 큰 수는 6210입니다. 0 또는 양의 정수가 담긴 배열 numbers가 매개변수로 주어질 때, 순서를 재배치하여 만들 수 있는 가장 큰 수를 문자열로 바꾸어 반환하도록 함수를 작성해주세요. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.
> 제한사항은 다음과 같습니다.
>
> - numbers의 길이는 1 이상 100,000 이하입니다.
> - numbers의 원소는 0 이상 1,000 이하입니다.

In [ ]:
def largest_number(numbers):
    # 각 숫자를 3번씩 이어 붙이고 sort해서 내림차순 정렬(numbers가 최대 3자리이므로)
    # 예) (입력) ['997', 9'] -> ['997997997', '999'] 비교 '999' > '997997997' -> (출력) ['9', '997']
    # numbers가 10^5 이기 때문에 다른 알고리즘의 경우 오래걸릴 수 있음
    # 단, 1000의 경우는 4자리이지만 영향을 미치지 않음
    sorted_numbers = sorted(numbers, key=lambda x: x * 3, reverse=True)

    largest_num = ''.join(sorted_numbers) # 정렬된 숫자를 합침

    # [0, 0, 0]과 같이 가장 앞자리가 0인 경우 0 출력
    if largest_num[0] == '0':
      return 0
    else:
      return largest_num

numbers = input("정수 리스트를 입력하세요 : ").split()
print(largest_number(numbers))

정수 리스트를 입력하세요 : 6 10 2
6210


Day 40
> 초 단위로 기록된 주식가격이 담긴 배열(리스트) prices가 매개변수로 주어질 때, 가격이 떨어지지 않은 기간은 몇 초인지를 배열(리스트)로 반환하도록 함수를 완성하세요. 이 함수가 제대로 동작하는지 테스트하는 프로그램을 작성하세요.
> 제한사항은 다음과 같습니다.
- prices의 각 가격은 1 이상 10,000 이하인 자연수입니다.
- prices의 길이는 2 이상 100,000 이하입니다.
>
>
>  입출력 예
- prices : [1, 2, 3, 2, 3]
- return : [4, 3, 1, 1, 0]
>
>입출력 예 설명
- 1초 시점의 ₩1은 끝까지 가격이 떨어지지 않았습니다.
- 2초 시점의 ₩2은 끝까지 가격이 떨어지지 않았습니다.
- 3초 시점의 ₩3은 1초뒤에 가격이 떨어집니다. 따라서 1초간 가격이 떨어지지 않은 것으로 봅니다.
- 4초 시점의 ₩2은 1초간 가격이 떨어지지 않았습니다.
- 5초 시점의 ₩3은 0초간 가격이 떨어지지 않았습니다

In [ ]:
def calc_price_time(prices):
  result = []

  for i, price in enumerate(prices):
    time = 0
    for j in range(i + 1, len(prices)):
      time += 1
      if price > prices[j]:
        break
    result.append(time)

  return result

prices = list(map(int, input("주식가격 리스트를 입력하세요 : ").split()))
print(calc_price_time(prices))

주식가격 리스트를 입력하세요 : 1 2 3 2 3
[4, 3, 1, 1, 0]


Day 41
> 리스트 안에 리스트가 중첩되어있는 아래와 같은 중첩 리스트(nested list)를
- [[1, 2, [3]], [4, [5, 6]], 7, [8, 9]]
- 다음과 같은 1차원 리스트로 평탄화하는 방법을 찾아보세요.
- [1, 2, 3, 4, 5, 6, 7, 8, 9]

In [9]:
def flatten_nested_list(nested_list):
  if isinstance(nested_list, list):
    result = []
    for item in nested_list:
      result.extend(flatten_nested_list(item))
    return result
  return [nested_list]

nested_list = [[1, 2, [3]], [4, [5, 6]], 7, [8, 9]]
print(flatten_nested_list(nested_list))

1
2
3
4
5
6
7
8
9
[1, 2, 3, 4, 5, 6, 7, 8, 9]
